### PR TITLE
feat: Autogenerate admin api key if not passed

### PR DESCRIPTION
**Pull request by Revolyssup**
### Description

Fixes # (issue)
This PR is part of the proposal which removes hardcoding of sensitive API keys by autogenerating them at either application level(APISIX) or deployment level.

## Notes

1. The admin api key is autogenerated if it is passed empty.
2. All the CLI tests now retrieve the key from config.yaml first before sending request to the API server, which is why the change in tests related to CLI.
3. If key is null or a non empty string, then it will not be changed.
4. This is not recommended at application level and recommended at infra level. For instance, the helm chart can autogenerate key. Therefore users are advised to pass their own unique key. For simpler local, single host deployments, APISIX can autogenerate the key which the users can read from config.yaml.

Related PRs:
https://github.com/apache/apisix-helm-chart/pull/740
https://github.com/apache/apisix-docker/pull/548

### Checklist

- [ ] I have explained the need for this PR and the problem it solves
- [ ] I have explained the changes or the new features added to this PR
- [ ] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: .github/workflows/source-install.yml [hunk 1]
```diff
@@ -88,8 +88,15 @@ jobs:
 
       - name: Test apisix
         run: |
-          curl http://127.0.0.1:9180/apisix/admin/routes/1 \
-            -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
+            wget https://github.com/mikefarah/yq/releases/download/3.4.1/yq_linux_amd64 -O /usr/bin/yq && sudo chmod +x /usr/bin/yq
+            get_admin_key() {
+            local admin_key=$(yq '.deployment.admin.admin_key[0].key' conf/config.yaml)
+            echo "$admin_key"
```
Comment (starsz) [inline .github/workflows/source-install.yml:94]:
```suggestion
            local admin_key=$(yq '.deployment.admin.admin_key[0].key' conf/config.yaml)
            echo "$admin_key"
```
```suggestion
                local admin_key=$(yq '.deployment.admin.admin_key[0].key' conf/config.yaml)
                echo "$admin_key"
```

File: Makefile [hunk 2]
```diff
@@ -129,8 +130,9 @@ deps: install-runtime
 	$(eval ENV_LUAROCKS_VER := $(shell $(ENV_LUAROCKS) --version | grep -E -o "luarocks [0-9]+."))
 	@if [ '$(ENV_LUAROCKS_VER)' = 'luarocks 3.' ]; then \
 		mkdir -p ~/.luarocks; \
-		$(ENV_LUAROCKS) config $(ENV_LUAROCKS_FLAG_LOCAL) variables.OPENSSL_LIBDIR $(addprefix $(ENV_OPENSSL_PREFIX), /lib); \
+		$(ENV_LUAROCKS) config $(ENV_LUAROCKS_FLAG_LOCAL) variables.OPENSSL_LIBDIR $(addprefix $(ENV_OPENSSL_PREFIX), /lib)  ; \
```
Comment (starsz) [inline Makefile:133]:
```suggestion
		$(ENV_LUAROCKS) config $(ENV_LUAROCKS_FLAG_LOCAL) variables.OPENSSL_LIBDIR $(addprefix $(ENV_OPENSSL_PREFIX), /lib); \
```

File: apisix/core/id.lua [hunk 3]
```diff
@@ -20,36 +20,38 @@
...
-local open             = io.open
-
-
+local try_read_attr = require("apisix.core.table").try_read_attr
+local profile = require("apisix.core.profile")
+local log = require("apisix.core.log")
+local uuid = require("resty.jit-uuid")
+local smatch = string.match
+local open = io.open
+local lyaml = require("lyaml")
```
Comment (starsz) [inline apisix/core/id.lua:29]: Please put this line after `local uuid = require("resty.jit-uuid")`

File: apisix/core/id.lua [hunk 4]
```diff
@@ -61,19 +63,88 @@ local function write_file(path, data)
...
+            end
+        end
+    end
+    return default_conf,changed
+end
 
 function _M.init()
+    local local_conf = fetch_local_conf()
+
+    local local_conf,changed = autogenerate_admin_key(local_conf)
```
Comment (starsz) [inline apisix/core/id.lua:131]: Only generate the admin key if the role is traditional and control plane.

File: apisix/core/id.lua [hunk 5]
```diff
@@ -61,19 +63,88 @@ local function write_file(path, data)
...
+    yaml = yaml:gsub(":%s*true%s*true", ": true"):gsub(":%s*false%s*true", ": false")
+
+    -- Replace *no_nil with true
+    yaml = yaml:gsub("&no_nil", "true")
+
+    -- Remove any occurrences of *no_nil
+    yaml = yaml:gsub(":%s*%*no_nil", ": true")
+
+    -- Remove duplicates for boolean values
+    yaml = yaml:gsub("true%s*true", "true"):gsub("false%s*false", "false")
```
Comment (starsz) [inline apisix/core/id.lua:98]: Why do we need those line ?

File: apisix/core/id.lua [hunk 6]
```diff
@@ -61,19 +63,88 @@ local function write_file(path, data)
...
+        and default_conf.deployment.admin.admin_key
+    if admin_keys and type(admin_keys) == "table" then
+        for i, admin_key in ipairs(admin_keys) do
+            if admin_key.role == "admin" and admin_key.key == "" then
+                changed = true
+                admin_keys[i].key = ""
+                for _ = 1, 32 do
+                    admin_keys[i].key = admin_keys[i].key ..
+                    string.char(math.random(65, 90) + math.random(0, 1) * 32)
+                end
```
Comment (starsz) [inline apisix/core/id.lua:121]: I thnk we can generate a uuid.

File: apisix/core/id.lua (hunk 3 above)
Comment (shreemaan-abhishek) [reply apisix/core/id.lua:29]: and beautify the code placing `=`s directly above each other.

File: apisix/core/id.lua [hunk 7]
```diff
@@ -20,36 +20,38 @@
...
 
-local _M = {version = 0.1}
-
+local _M = { version = 0.1 }
 
 local function rtrim(str)
     return smatch(str, "^(.-)%s*$")
 end
 
-
```
Comment (shreemaan-abhishek) [inline apisix/core/id.lua:40]: two blank lines are needed between functions, please check in other places as well.

File: apisix/core/id.lua [hunk 8]
```diff
@@ -20,36 +20,38 @@
...
 local function read_file(path)
     local file = open(path, "rb") -- r read mode and b binary mode
     if not file then
         return nil
     end
 
-    local content = file:read("*a")  -- *a or *all reads the whole file
+    local content = file:read("*a") -- *a or *all reads the whole file
```
Comment (shreemaan-abhishek) [inline apisix/core/id.lua:50]: better no not introduce changes like this, as this is already a big PR. Plus, a PR should only contain changes meaningful to the PR title.

File: apisix/core/id.lua (hunk 4 above)
Comment (shreemaan-abhishek) [inline apisix/core/id.lua:131]:
```suggestion
    local local_conf,changed = autogenerate_admin_key(local_conf)
```
```suggestion
    local local_conf, changed = autogenerate_admin_key(local_conf)
```

File: ci/centos7-ci.sh [hunk 9]
```diff
@@ -24,11 +24,12 @@ install_dependencies() {
 
     # install build & runtime deps
     yum install -y wget tar gcc gcc-c++ automake autoconf libtool make unzip patch \
-        git sudo openldap-devel which ca-certificates \
+        git sudo openldap-devel which ca-certificates\
```
Comment (shreemaan-abhishek) [inline ci/centos7-ci.sh:27]: please remove such unnecessary changes.

File: conf/config-default.yaml [hunk 10]
```diff
@@ -170,7 +170,8 @@ nginx_config:                     # Config for render the template to generate n
   stream:
     enable_access_log: false                 # Enable stream proxy access logging.
     access_log: logs/access_stream.log       # Location of the stream access log.
-    access_log_format: "$remote_addr [$time_local] $protocol $status $bytes_sent $bytes_received $session_time" # Customize log format: http://nginx.org/en/docs/varindex.html
+    access_log_format: |
+      "$remote_addr [$time_local] $protocol $status $bytes_sent $bytes_received $session_time" # Customize log format: http://nginx.org/en/docs/varindex.html
```
Comment (shreemaan-abhishek) [inline conf/config-default.yaml:174]: due to this change, are you sure the comment won't be parsed as a part of the string?

File: apisix/core/id.lua (hunk 5 above)
Comment (Revolyssup) [reply apisix/core/id.lua:98]: The lyaml library doesnt do null parsing properly. It converts NULL into`[]:` producing invalid yaml. And when I use the workaround of replacing null [here](https://github.com/apache/apisix/pull/11080/files/26a948aa6366ec39de33d875e66c0a819d5526ea#diff-ab3c55c5825b9d6c1f2a2e296c658ea10092998b5a53feca14fd3988bcf8209aR80), it messes up boolean values so these lines fix them back. I will refactor it to make it a little concise.

File: apisix/core/id.lua (hunk 4 above)
Comment (Revolyssup) [reply apisix/core/id.lua]: okay

File: apisix/core/id.lua (hunk 5 above)
Comment (Revolyssup) [reply apisix/core/id.lua]: I removed the unnecessary lines.

File: apisix/core/id.lua (hunk 6 above)
Comment (Revolyssup) [reply apisix/core/id.lua:114]: I am simply generating an alphanumeric and I think this can be done without using any external module for generating UUIDs. This is straightforward and doesnt require any other module. Which function or module do you suggest to generate a uuid?

File: docs/en/latest/admin-api.md [hunk 11]
```diff
@@ -122,24 +122,34 @@ By default, the Admin API checks for references between resources and will refus
...
+
+```bash
+ admin_key=$(yq '.deployment.admin.admin_key[0].key' conf/config.yaml | sed 's/"//g')
+```
+
+:::
+
 ```bash
-$ curl http://127.0.0.1:9180/apisix/admin/upstreams/1 -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '{
+$ curl http://127.0.0.1:9180/apisix/admin/upstreams/1 -H "X-API-KEY: $admin_key" -X PUT -d '{
```
Comment (kayx23) [inline docs/en/latest/admin-api.md:136]: Wouldn't this be interpreted as string literal? Should it be `${admin_key}` instead? Because I saw you save the admin key to an env var as well

File: docs/en/latest/admin-api.md [hunk 12]
```diff
@@ -122,24 +122,34 @@ By default, the Admin API checks for references between resources and will refus
 
 You can make a force deletion by adding the request argument `force=true` to the delete request, for example:
 
+:::note
+
+You can get the get the admin_key from config.yaml like this.
```
Comment (kayx23) [inline docs/en/latest/admin-api.md:127]:
```suggestion
You can fetch the `admin_key` from `config.yaml` and save to an environment variable with the following command:
```

Comment (kayx23) [reply docs/en/latest/admin-api.md:127]: Please update all other instances as well if this suggestion is accepted.

File: docs/zh/latest/plugins/proxy-mirror.md [hunk 13]
```diff
@@ -49,9 +49,19 @@ description: 本文介绍了 Apache APISIX proxy-mirror 插件的相关操作，
 
 以下示例展示了如何在指定路由上启用 `proxy-mirror` 插件：
 
+:::note
+
+您可以像这样从 config.yaml 中获取 admin_key。
```
Comment (kayx23) [inline docs/zh/latest/plugins/proxy-mirror.md:54]:
```suggestion
您可以这样从 `config.yaml` 中获取 `admin_key` 并存入环境变量：
```

**Review by kayx23 — CHANGES_REQUESTED** [review]
a few suggestions for the docs

File: apisix/core/id.lua (hunk 6 above)
Comment (starsz) [reply apisix/core/id.lua:114]: > Which function or module do you suggest to generate a uuid?

You can review the `request-id` plugin.

File: apisix/core/id.lua [hunk 14]
```diff
@@ -62,18 +67,78 @@ local function write_file(path, data)
...
+local function autogenerate_admin_key(default_conf)
+    local changed = false
+   -- Check if deployment.role is either traditional or control_plane
+    local deployment_role = default_conf.deployment and default_conf.deployment.role
+    if deployment_role and (deployment_role == "traditional" or
+       deployment_role == "control_plane") then
+        -- Check if deployment.admin.admin_key is not nil and it's an empty string
+        local admin_keys = default_conf.deployment
+            and default_conf.deployment.admin
+            and default_conf.deployment.admin.admin_key
```
Comment (starsz) [inline apisix/core/id.lua:104]: I think you can try `core.table.try_read_attr` to improve your code.

File: apisix/core/id.lua [hunk 15]
```diff
@@ -62,18 +67,78 @@ local function write_file(path, data)
...
 function _M.init()
+    local local_conf = fetch_local_conf()
+
+    local local_conf, changed = autogenerate_admin_key(local_conf)
+    if changed then
+        local yaml_conf = generate_yaml(local_conf)
+        local local_conf_path = profile:yaml_path("config")
+        local ok, err = write_file(local_conf_path, yaml_conf)
+        if not ok then
+            log.error(err)
```
Comment (starsz) [inline apisix/core/id.lua:131]: I think you should panic if we can't write file.

File: apisix/core/id.lua (hunk 6 above)
Comment (moonming) [reply apisix/core/id.lua:114]: https://github.com/apache/apisix/pull/11080/files#diff-ab3c55c5825b9d6c1f2a2e296c658ea10092998b5a53feca14fd3988bcf8209aR26 here is the uuid lib if you decide to use uuid

File: conf/config-default.yaml (hunk 10 above)
Comment (moonming) [reply conf/config-default.yaml:174]: This change has nothing to do with this modification. If you think it is necessary, it is recommended to open a new PR to deal with it.

File: conf/config-default.yaml [hunk 16]
```diff
@@ -208,7 +209,8 @@ nginx_config:                     # Config for render the template to generate n
     enable_access_log: true             # Enable HTTP proxy access logging.
     access_log: logs/access.log         # Location of the access log.
     access_log_buffer: 16384            # buffer size of access log.
-    access_log_format: "$remote_addr - $remote_user [$time_local] $http_host \"$request\" $status $body_bytes_sent $request_time \"$http_referer\" \"$http_user_agent\" $upstream_addr $upstream_status $upstream_response_time \"$upstream_scheme://$upstream_host$upstream_uri\""
+    access_log_format: |
+      "$remote_addr - $remote_user [$time_local] $http_host \"$request\" $status $body_bytes_sent $request_time \"$http_referer\" \"$http_user_agent\" $upstream_addr $upstream_status $upstream_response_time \"$upstream_scheme://$upstream_host$upstream_uri\""
```
Comment (moonming) [inline conf/config-default.yaml:213]: ditto

Comment (Revolyssup) [reply conf/config-default.yaml:213]: The lyaml.dump function which is being used here to convert lua table into yaml, messes up the formatting and adds a new line automatically. Using `|` operator avoids this issue by allowing for multiline string

File: docs/zh/latest/plugins/proxy-mirror.md (hunk 13 above)
Comment (Revolyssup) [reply docs/zh/latest/plugins/proxy-mirror.md:54]: done

**Review by moonming — DISMISSED** [review]

File: apisix/core/id.lua (hunk 6 above)
Comment (shreemaan-abhishek) [reply apisix/core/id.lua:113]: @Revolyssup please address these comments.

File: apisix/core/id.lua [hunk 17]
```diff
@@ -20,17 +20,23 @@
...
 
 local fetch_local_conf = require("apisix.core.config_local").local_conf
-local try_read_attr    = require("apisix.core.table").try_read_attr
-local log              = require("apisix.core.log")
-local uuid             = require('resty.jit-uuid')
-local smatch           = string.match
-local open             = io.open
-
-
+local try_read_attr = require("apisix.core.table").try_read_attr
```
Comment (shreemaan-abhishek) [inline apisix/core/id.lua:23]: please maintain code beautification (equals on top of each other)

File: apisix/core/id.lua [hunk 18]
```diff
@@ -20,17 +20,23 @@
...
+local ipairs = ipairs
+local string = string
+local math = math
 local prefix = ngx.config.prefix()
 local apisix_uid
+local pairs = pairs
+local ngx_exit = ngx.exit
 
-local _M = {version = 0.1}
+local _M = { version = 0.1 }
```
Comment (shreemaan-abhishek) [inline apisix/core/id.lua:39]: please revert this change.

File: t/fuzzing/simple_http.py [hunk 19]
```diff
@@ -41,10 +49,18 @@ def create_route():
         }
     })
     conn = connect_admin()
+    key = get_admin_key_from_yaml('conf/config.yaml')
+    if key is None:
+        print("Key not found in the YAML file.")
+        return
+    key = key.replace('"', '')
+    print("the key is",key)
```
Comment (shreemaan-abhishek) [inline t/fuzzing/simple_http.py:57]:
```suggestion
    print("the key is", key)
```

File: .github/workflows/source-install.yml [hunk 20]
```diff
@@ -88,8 +88,15 @@ jobs:
 
       - name: Test apisix
         run: |
-          curl http://127.0.0.1:9180/apisix/admin/routes/1 \
-            -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
+            wget https://github.com/mikefarah/yq/releases/download/3.4.1/yq_linux_amd64 -O /usr/bin/yq && sudo chmod +x /usr/bin/yq
```
Comment (shreemaan-abhishek) [inline .github/workflows/source-install.yml:91]: this block has 4 indents, where as everywhere in this file there are 2 indents.

File: .github/workflows/source-install.yml [hunk 21]
```diff
@@ -99,15 +106,15 @@ jobs:
...
-          result_code=`curl -I -m 10 -o /dev/null -s -w %{http_code} http://127.0.0.1:9080/get`
-          if [[ $result_code -ne 200 ]]; then
-              printf "result_code: %s\n" "$result_code"
-              echo "===============access.log==============="
-              cat logs/access.log
-              echo "===============error.log==============="
-              cat logs/error.log
-              exit 125
-          fi
+            result_code=`curl -I -m 10 -o /dev/null -s -w %{http_code} http://127.0.0.1:9080/get`
```
Comment (shreemaan-abhishek) [inline .github/workflows/source-install.yml:109]: this change is not needed.

File: conf/config-default.yaml (hunk 10 above)
Comment (shreemaan-abhishek) [reply conf/config-default.yaml:174]: @Revolyssup please address this comment.

**Comment by Revolyssup** [comment]
@shreemaan-abhishek Comment addressed here https://github.com/apache/apisix/pull/11080#discussion_r1546263066

File: .github/workflows/source-install.yml (hunk 21 above)
Comment (Revolyssup) [reply .github/workflows/source-install.yml:109]: fixed

File: .github/workflows/source-install.yml (hunk 20 above)
Comment (Revolyssup) [reply .github/workflows/source-install.yml:91]: fixed

File: apisix/core/id.lua (hunk 17 above)
Comment (Revolyssup) [reply apisix/core/id.lua:23]: done

File: docs/en/latest/plugins/api-breaker.md [hunk 22]
```diff
@@ -57,9 +57,18 @@ In an unhealthy state, if the Upstream service responds with a status code from
 
 The example below shows how you can configure the Plugin on a specific Route:
 
+:::note
+You can fetch the `admin_key` from `config.yaml` and save to an environment variable with the following command:
+
+```bash
+ admin_key=$(yq '.deployment.admin.admin_key[0].key' conf/config.yaml | sed 's/"//g')
```
Comment (kayx23) [inline docs/en/latest/plugins/api-breaker.md:64]:
```suggestion
admin_key=$(yq '.deployment.admin.admin_key[0].key' conf/config.yaml | sed 's/"//g')
```
Remove extra space. Please replace all.

**Review by kayx23 — COMMENTED** [review]
Other doc changes LGTM.

File: apisix/core/id.lua [hunk 23]
```diff
@@ -20,23 +20,22 @@
...
-local lyaml = require("lyaml")
-local smatch = string.match
-local open = io.open
-local type = type
-local ipairs = ipairs
-local string = string
-local math = math
-local prefix = ngx.config.prefix()
+local try_read_attr    = require("apisix.core.table").try_read_attr
+local profile           = require("apisix.core.profile")
```
Comment (shreemaan-abhishek) [inline apisix/core/id.lua:24]: ?

File: apisix/core/id.lua [hunk 24]
```diff
@@ -20,23 +20,22 @@
...
+local log              = require("apisix.core.log")
+local uuid             = require("resty.jit-uuid")
+local lyaml            = require("lyaml")
+local smatch           = string.match
+local open             = io.open
+local type             = type
+local ipairs           = ipairs
+local string           = string
+local math             = math
+local prefix            = ngx.config.prefix()
```
Comment (shreemaan-abhishek) [inline apisix/core/id.lua:34]: ?

File: apisix/core/id.lua [hunk 25]
```diff
@@ -20,23 +20,22 @@
...
+local pairs            = pairs
+local ngx_exit         = ngx.exit
 local apisix_uid
-local pairs = pairs
-local ngx_exit = ngx.exit
-
-local _M = { version = 0.1 }
+local _M = {version = 0.1}
```
Comment (shreemaan-abhishek) [inline apisix/core/id.lua:38]:
```suggestion

local _M = {version = 0.1}
```

**Review by shreemaan-abhishek — APPROVED** [review]

**Review by kayx23 — APPROVED** [review]
doc changes LGTM
